pkg/apis/build/v1alpha1/build_types.go

```diff
@@ -137,20 +149,28 @@ type ArgumentSpec struct {
 // SourceSpec defines the input to the Build
 type SourceSpec struct {
 	// Git represents source in a Git repository.
+	// +optional
 	Git *GitSourceSpec `json:"git,omitempty"`
 
 	// GCS represents source in Google Cloud Storage.
+	// +optional
 	GCS *GCSSourceSpec `json:"gcs,omitempty"`
 
 	// Custom indicates that source should be retrieved using a custom
 	// process defined in a container invocation.
+	// +optional
 	Custom *corev1.Container `json:"custom,omitempty"`
 
 	// SubPath specifies a path within the fetched source which should be
 	// built. This option makes parent directories *inaccessible* to the
 	// build steps. (The specific source type may, in fact, not even fetch
 	// files not in the SubPath.)
+	// +optional
 	SubPath string `json:"subPath,omitempty"`
+
+	// Name is the path under which source will be mounted.
+	// +optional
+	Name string `json:"name,omitempty"`
```

Comment:
Since this gets appended into the container name, we should note the restrictions on allowed characters:
* Must be a basename (no /)
* Must be a [valid DNS name](https://stackoverflow.com/questions/27791913/what-characters-are-allowed-in-kubernetes-port-and-container-names) (only alphanumeric characters, no _ or -)

Also would "Path", "TargetSubPath" etc be a better name here?

Reply:
- If Path or TagetPath is used then it should allow `/` character.  
- I thought `Name` would be a good option because its unique identity of source.

I do like the TargetPath feature but I think it is separate feature and I am not sure about repurposing `Name` to be targetPath. I understand the use case and I think it is completely valid

Reply:
Once/if we have TargetPath what is the purpose of Name?

I think we should consider replacing Name with TargetPath as this seems like an important feature because it is an essential feature for go projects. And go is an important language for knative and kubernetes.

Including `TargetPath` support seems more important than the `WorkDir` (cloning to /workspace) conversation below.

The workaround here would be to add a step which moves each source from /workspace/foo to /workspace/go/src/whatever.domain/path/to/package, which may be complicated to extract from "foo" (especially since foo can't contain most obvious separate chars)

Reply:
You have valid points. 
- I am misusing `Name` as TargetPath in this PR.  I will rename the variable to `TargetPath`. Question: Is  there value in adding validation to check if multiple sources don't have same `TargetPath` or is that too much validation in build spec ?

- I think there is still value in having `Name` field but the purpose is purely for logging, generating init container names. Init containers with meaningful names made my life easy when debugging. Make the field optional for user so less things to declare I guess. WDYT?